File: shcore/src/printf.rs
//! The `printf` formatting engine — pure byte-in/byte-out, no OS.
//!
//! This is the SINGLE home for shell `printf` formatting. The `printf` builtin
//! (`exec.rs`) and the `/bin/printf` tool (`programs/`) both call [`render`], so the
//! intricate format/escape handling exists once and cannot drift between them — the
//! "carry-both" pair shares code instead of a hand-sync comment (B2 in spirit).
//!
//! Conversions: `%s %b %c %d %i %u %o %x %X %%` with flags (`- + space 0 #`), width,
//! and `.precision` (each may be `*` from an argument), plus C backslash escapes.
//! Unknown conversions (e.g. `%f`) are emitted verbatim. `\n` is a literal LF — the
//! terminal adds CR via ONLCR, so files and pipes stay LF.

use alloc::vec::Vec;

/// Format `fmt` against `args`, reusing the format string while it keeps consuming
/// arguments (POSIX). Returns the rendered bytes and whether a numeric conversion
/// error occurred (the caller maps that to a nonzero exit). `\c` stops output.
pub fn render(fmt: &[u8], args: &[&[u8]]) -> (Vec<u8>, bool) {
    let mut out: Vec<u8> = Vec::new();
    let mut ai = 0usize;
    let mut err = false;
    loop {
        let before = ai;
        let stop = render_once(fmt, args, &mut ai, &mut out, &mut err);
        if stop {
            break;
        }
        // Reuse the format only while it keeps consuming arguments.
        if ai >= args.len() || ai == before {
            break;
        }
    }
    (out, err)
}

fn push_radix(out: &mut Vec<u8>, mut v: u64, radix: u64, upper: bool) {
    const LOWER: &[u8; 16] = b"0123456789abcdef";
    const UPPER: &[u8; 16] = b"0123456789ABCDEF";
    let digits = if upper { UPPER } else { LOWER };
    let mut tmp = [0u8; 64];
    let mut i = tmp.len();
    if v == 0 {
        i -= 1;
        tmp[i] = b'0';
    }
    while v > 0 {
        i -= 1;
        tmp[i] = digits[(v % radix) as usize];
        v /= radix;
    }
    out.extend_from_slice(&tmp[i..]);
}

fn hexval(c: u8) -> Option<u32> {
    match c {
        b'0'..=b'9' => Some((c - b'0') as u32),
        b'a'..=b'f' => Some((c - b'a' + 10) as u32),
        b'A'..=b'F' => Some((c - b'A' + 10) as u32),
        _ => None,
    }
}

fn push_utf8(out: &mut Vec<u8>, cp: u32) {
    if cp < 0x80 {
        out.push(cp as u8);
    } else if cp < 0x800 {
        out.push(0xC0 | (cp >> 6) as u8);
        out.push(0x80 | (cp & 0x3f) as u8);
    } else if cp < 0x10000 {
        out.push(0xE0 | (cp >> 12) as u8);
        out.push(0x80 | ((cp >> 6) & 0x3f) as u8);
        out.push(0x80 | (cp & 0x3f) as u8);
    } else {
        out.push(0xF0 | (cp >> 18) as u8);
        out.push(0x80 | ((cp >> 12) & 0x3f) as u8);
        out.push(0x80 | ((cp >> 6) & 0x3f) as u8);
        out.push(0x80 | (cp & 0x3f) as u8);
    }
}

/// Decode one backslash escape at `s[i]` (where `s[i] == b'\\'`), pushing its
/// byte(s) to `out`. Returns `(next_index, stop)`; `stop` is true for `\c`.
pub fn escape_at(s: &[u8], i: usize, out: &mut Vec<u8>) -> (usize, bool) {
    if i + 1 >= s.len() {
        out.push(b'\\');
        return (i + 1, false);
    }
    let n = s[i + 1];
    match n {
        b'a' => (i + 2, false_push(out, 7)),
        b'b' => (i + 2, false_push(out, 8)),
        b'f' => (i + 2, false_push(out, 12)),
        b'n' => (i + 2, false_push(out, b'\n')),
        b'r' => (i + 2, false_push(out, b'\r')),
        b't' => (i + 2, false_push(out, b'\t')),
        b'v' => (i + 2, false_push(out, 11)),
        b'\\' => (i + 2, false_push(out, b'\\')),
        b'c' => (i + 2, true),
        b'x' => {
            let mut j = i + 2;
            let mut val = 0u32;
            let mut cnt = 0;
            while j < s.len() && cnt < 2 {
                match hexval(s[j]) {
                    Some(h) => {
                        val = val * 16 + h;
                        j += 1;
                        cnt += 1;
                    }
                    None => break,
                }
            }
            if cnt == 0 {
                out.push(b'\\');
                out.push(b'x');
                (i + 2, false)
            } else {
                out.push((val & 0xff) as u8);
                (j, false)
            }
        }
        b'u' | b'U' => {
            let max = if n == b'u' { 4 } else { 8 };
            let mut j = i + 2;
            let mut val = 0u32;
            let mut cnt = 0;
            while j < s.len() && cnt < max {
                match hexval(s[j]) {
                    Some(h) => {
                        val = val * 16 + h;
                        j += 1;
                        cnt += 1;
                    }
                    None => break,
                }
            }
            if cnt == 0 {
                out.push(b'\\');
                out.push(n);
                (i + 2, false)
            } else {
                push_utf8(out, val);
                (j, false)
            }
        }
        b'0' => {
            // \0NNN — up to three octal digits after the 0.
            let mut j = i + 2;
            let mut val = 0u32;
            let mut cnt = 0;
            while j < s.len() && cnt < 3 && (b'0'..=b'7').contains(&s[j]) {
                val = val * 8 + (s[j] - b'0') as u32;
                j += 1;
                cnt += 1;
            }
            out.push((val & 0xff) as u8);
            (j, false)
        }
        b'1'..=b'7' => {
            // \NNN — this digit plus up to two more octal digits.
            let mut j = i + 1;
            let mut val = 0u32;
            let mut cnt = 0;
            while j < s.len() && cnt < 3 && (b'0'..=b'7').contains(&s[j]) {
                val = val * 8 + (s[j] - b'0') as u32;
                j += 1;
                cnt += 1;
            }
            out.push((val & 0xff) as u8);
            (j, false)
        }
        other => {
            out.push(b'\\');
            out.push(other);
            (i + 2, false)
        }
    }
}

/// Push `b` and return `false` (a helper so escape arms stay one-liners).
fn false_push(out: &mut Vec<u8>, b: u8) -> bool {
    out.push(b);
    false
}

/// Parse the integer value of a numeric argument: a leading `'`/`"` yields the
/// next byte's value (POSIX); otherwise an optionally-signed decimal. Sets
/// `*err` if the argument has no digits or trailing junk (GNU diagnoses but
/// still uses the parsed prefix / 0).
fn arg_num(a: &[u8], err: &mut bool) -> i64 {
    if a.is_empty() {
        return 0;
    }
    if a[0] == b'\'' || a[0] == b'"' {
        return a.get(1).copied().unwrap_or(0) as i64;
    }
    let neg = a[0] == b'-';
    let mut i = if neg || a[0] == b'+' { 1 } else { 0 };
    let start = i;
    let mut v: i64 = 0;
    while i < a.len() && a[i].is_ascii_digit() {
        v = v.saturating_mul(10).saturating_add((a[i] - b'0') as i64);
        i += 1;
    }
    if i == start || i != a.len() {
        *err = true;
    }
    if neg {
        -v
    } else {
        v
    }
}

/// Emit a numeric body (`digits`, preceded by `sign`, which may also be a `0x`
/// prefix) honoring precision (minimum digit count), width, and the `-`/`0` flags.
fn emit_num(
    out: &mut Vec<u8>,
    sign: &[u8],
    digits: &[u8],
    width: usize,
    prec: Option<usize>,
    left: bool,
    zero: bool,
) {
    let mut d: Vec<u8> = Vec::new();
    if let Some(p) = prec {
        if p > digits.len() {
            d.resize(p - digits.len(), b'0');
        }
    }
    d.extend_from_slice(digits);
    let content = sign.len() + d.len();
    let pad = width.saturating_sub(content);
    if left {
        out.extend_from_slice(sign);
        out.extend_from_slice(&d);
        out.resize(out.len() + pad, b' ');
    } else if zero && prec.is_none() {
        out.extend_from_slice(sign);
        out.resize(out.len() + pad, b'0');
        out.extend_from_slice(&d);
    } else {
        out.resize(out.len() + pad, b' ');
        out.extend_from_slice(sign);
        out.extend_from_slice(&d);
    }
}

/// Emit a string body honoring width and the `-` flag (the `0` flag never
/// applies to strings).
fn emit_str(out: &mut Vec<u8>, body: &[u8], width: usize, left: bool) {
    let pad = width.saturating_sub(body.len());
    if left {
        out.extend_from_slice(body);
        out.resize(out.len() + pad, b' ');
    } else {
        out.resize(out.len() + pad, b' ');
        out.extend_from_slice(body);
    }
}

/// Render `fmt` once, consuming arguments from `args` starting at `*ai`. Returns
/// `true` if `\c` was seen (caller stops entirely). `*err` records a numeric
/// conversion error.
fn render_once(
    fmt: &[u8],
    args: &[&[u8]],
    ai: &mut usize,
    out: &mut Vec<u8>,
    err: &mut bool,
) -> bool {
    let mut i = 0;
    while i < fmt.len() {
        let c = fmt[i];
        if c == b'\\' {
            let (ni, stop) = escape_at(fmt, i, out);
            i = ni;
            if stop {
                return true;
            }
            continue;
        }
        if c != b'%' || i + 1 >= fmt.len() {
            out.push(c);
            i += 1;
            continue;
        }
        // A `%` directive.
        i += 1;
        if fmt[i] == b'%' {
            out.push(b'%');
            i += 1;
            continue;
        }
        let spec_start = i;
        // Flags.
        let (mut left, mut plus, mut space, mut zero, mut hash) =
            (false, false, false, false, false);
        while i < fmt.len() {
            match fmt[i] {
                b'-' => left = true,
                b'+' => plus = true,
                b' ' => space = true,
                b'0' => zero = true,
                b'#' => hash = true,
                b'\'' => {} // grouping flag: ignored (C locale)
                _ => break,
            }
            i += 1;
        }
        // Width (number or `*`).
        let mut width: usize = 0;
        if i < fmt.len() && fmt[i] == b'*' {
            i += 1;
            let w = arg_num(args.get(*ai).copied().unwrap_or(b""), err);
            *ai += 1;
            if w < 0 {
                left = true;
                width = (-w) as usize;
            } else {
                width = w as usize;
            }
        } else {
            while i < fmt.len() && fmt[i].is_ascii_digit() {
                width = width * 10 + (fmt[i] - b'0') as usize;
                i += 1;
            }
        }
        // Precision (`.` then number or `*`).
        let mut prec: Option<usize> = None;
        if i < fmt.len() && fmt[i] == b'.' {
            i += 1;
            if i < fmt.len() && fmt[i] == b'*' {
                i += 1;
                let p = arg_num(args.get(*ai).copied().unwrap_or(b""), err);
                *ai += 1;
                prec = if p < 0 { None } else { Some(p as usize) };
            } else {
                let mut p = 0usize;
                while i < fmt.len() && fmt[i].is_ascii_digit() {
                    p = p * 10 + (fmt[i] - b'0') as usize;
                    i += 1;
                }
                prec = Some(p);
            }
        }
        // Skip C length modifiers (no effect here).
        while i < fmt.len() && matches!(fmt[i], b'l' | b'h' | b'L' | b'q' | b'j' | b'z' | b't') {
            i += 1;
        }
        if i >= fmt.len() {
            // Trailing incomplete directive: emit verbatim.
            out.push(b'%');
            out.extend_from_slice(&fmt[spec_start..i]);
            break;
        }
        let conv = fmt[i];
        i += 1;
        let arg = args.get(*ai).copied().unwrap_or(b"");
        match conv {
            b's' => {
                let body = match prec {
                    Some(p) => &arg[..p.min(arg.len())],
                    None => arg,
                };
                emit_str(out, body, width, left);
                *ai += 1;
            }
            b'b' => {
                let mut decoded: Vec<u8> = Vec::new();
                let mut j = 0;
                let mut stop = false;
                while j < arg.len() {
                    if arg[j] == b'\\' {
                        let (nj, st) = escape_at(arg, j, &mut decoded);
                        j = nj;
                        if st {
                            stop = true;
                            break;
                        }
                    } else {
                        decoded.push(arg[j]);
                        j += 1;
                    }
                }
                let body = match prec {
                    Some(p) => &decoded[..p.min(decoded.len())],
                    None => &decoded[..],
                };
                emit_str(out, body, width, left);
                *ai += 1;
                if stop {
                    return true;
                }
            }
            b'c' => {
                let one = &arg[..arg.len().min(1)];
                emit_str(out, one, width, left);
                *ai += 1;
            }
            b'd' | b'i' => {
                let v = arg_num(arg, err);
                let mut digits: Vec<u8> = Vec::new();
                if !(prec == Some(0) && v == 0) {
                    let mag = (v as i128).unsigned_abs() as u64;
                    push_radix(&mut digits, mag, 10, false);
                }
                let sign: &[u8] = if v < 0 {
                    b"-"
                } else if plus {
                    b"+"
                } else if space {
                    b" "
                } else {
                    b""
                };
                emit_num(out, sign, &digits, width, prec, left, zero);
                *ai += 1;
            }
            b'u' | b'o' | b'x' | b'X' => {
                let v = arg_num(arg, err) as u64;
                let mut digits: Vec<u8> = Vec::new();
                let radix = match conv {
                    b'o' => 8,
                    b'u' => 10,
                    _ => 16,
                };
                if !(prec == Some(0) && v == 0) {
                    push_radix(&mut digits, v, radix, conv == b'X');
                }
                let mut sign: Vec<u8> = Vec::new();
                if hash && v != 0 {
                    match conv {
                        b'o' => {
                            if digits.first() != Some(&b'0') {
                                sign.push(b'0');
                            }
                        }
                        b'x' => sign.extend_from_slice(b"0x"),
                        b'X' => sign.extend_from_slice(b"0X"),
                        _ => {}
                    }
                }
                emit_num(out, &sign, &digits, width, prec, left, zero);
                *ai += 1;
            }
            _ => {
                // Unknown conversion (e.g. %f): emit verbatim.
                out.push(b'%');
                out.extend_from_slice(&fmt[spec_start..i]);
            }
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::vec;

    fn r(fmt: &str, args: &[&str]) -> alloc::string::String {
        let a: Vec<&[u8]> = args.iter().map(|s| s.as_bytes()).collect();
        let (out, _err) = render(fmt.as_bytes(), &a);
        alloc::string::String::from_utf8_lossy(&out).into_owned()
    }

    #[test]
    fn integers_width_flags() {
        assert_eq!(r("%d", &["42"]), "42");
        assert_eq!(r("%5d", &["42"]), "   42");
        assert_eq!(r("%-5d", &["42"]), "42   ");
        assert_eq!(r("%05d", &["42"]), "00042");
        assert_eq!(r("%+d", &["42"]), "+42");
        assert_eq!(r("%d", &["-7"]), "-7");
        assert_eq!(r("%d", &[]), "0"); // missing arg ⇒ 0
    }

    #[test]
    fn radixes_and_hash() {
        assert_eq!(r("%x", &["255"]), "ff");
        assert_eq!(r("%X", &["255"]), "FF");
        assert_eq!(r("%#x", &["255"]), "0xff");
        assert_eq!(r("%o", &["8"]), "10");
        assert_eq!(r("%u", &["10"]), "10");
    }

    #[test]
    fn strings_and_precision() {
        assert_eq!(r("%s", &["hi"]), "hi");
        assert_eq!(r("%.2s", &["hello"]), "he");
        assert_eq!(r("%5s", &["hi"]), "   hi");
        assert_eq!(r("%-5s|", &["hi"]), "hi   |");
        assert_eq!(r("%c", &["abc"]), "a");
    }

    #[test]
    fn escapes_and_literal_percent() {
        assert_eq!(r("a\\tb\\n", &[]), "a\tb\n");
        assert_eq!(r("100%%", &[]), "100%");
        // \c stops all further output.
        assert_eq!(r("ab\\cde", &[]), "ab");
    }

    #[test]
    fn format_is_reused_while_args_remain() {
        assert_eq!(r("%d ", &["1", "2", "3"]), "1 2 3 ");
    }

    #[test]
    fn unknown_conversion_is_verbatim() {
        // %f is not implemented; it passes through unchanged.
        assert_eq!(r("[%3.2f]", &["1.5"]), "[%3.2f]");
    }

    #[test]
    fn star_width_from_arg() {
        assert_eq!(r("%*d", &["4", "7"]), "   7");
        let _ = vec![0u8]; // ensure alloc::vec import used in no_std test cfg
    }
}

File: memcontainers/shcore/src/printf.zig
//! printf formatting engine: pure byte-in/byte-out, no OS.

const std = @import("std");

pub const RenderResult = struct {
    bytes: []u8,
    had_error: bool = false,
};

pub fn render(allocator: std.mem.Allocator, fmt: []const u8, args: []const []const u8) !RenderResult {
    var out = std.ArrayList(u8).empty;
    var ai: usize = 0;
    var err = false;
    while (true) {
        const before = ai;
        const stop = try renderOnce(allocator, fmt, args, &ai, &out, &err);
        if (stop) break;
        if (ai >= args.len or ai == before) break;
    }
    return .{ .bytes = try out.toOwnedSlice(allocator), .had_error = err };
}

fn renderOnce(
    allocator: std.mem.Allocator,
    fmt: []const u8,
    args: []const []const u8,
    ai: *usize,
    out: *std.ArrayList(u8),
    err: *bool,
) !bool {
    var i: usize = 0;
    while (i < fmt.len) {
        const c = fmt[i];
        if (c == '\\') {
            const escaped = try escapeAt(allocator, fmt, i, out);
            i = escaped.next;
            if (escaped.stop) return true;
            continue;
        }
        if (c != '%' or i + 1 >= fmt.len) {
            try out.append(allocator, c);
            i += 1;
            continue;
        }

        i += 1;
        if (fmt[i] == '%') {
            try out.append(allocator, '%');
            i += 1;
            continue;
        }

        const spec_start = i;
        var left = false;
        var plus = false;
        var space = false;
        var zero = false;
        var hash = false;
        while (i < fmt.len) : (i += 1) {
            switch (fmt[i]) {
                '-' => left = true,
                '+' => plus = true,
                ' ' => space = true,
                '0' => zero = true,
                '#' => hash = true,
                '\'' => {},
                else => break,
            }
        }

        var width: usize = 0;
        if (i < fmt.len and fmt[i] == '*') {
            i += 1;
            const w = argNum(argAt(args, ai.*), err);
            ai.* += 1;
            if (w < 0) {
                left = true;
                width = @intCast(-w);
            } else {
                width = @intCast(w);
            }
        } else {
            while (i < fmt.len and std.ascii.isDigit(fmt[i])) : (i += 1) {
                width = width * 10 + (fmt[i] - '0');
            }
        }

        var precision: ?usize = null;
        if (i < fmt.len and fmt[i] == '.') {
            i += 1;
            if (i < fmt.len and fmt[i] == '*') {
                i += 1;
                const p = argNum(argAt(args, ai.*), err);
                ai.* += 1;
                precision = if (p < 0) null else @as(usize, @intCast(p));
            } else {
                var p: usize = 0;
                while (i < fmt.len and std.ascii.isDigit(fmt[i])) : (i += 1) {
                    p = p * 10 + (fmt[i] - '0');
                }
                precision = p;
            }
        }

        while (i < fmt.len and isLengthModifier(fmt[i])) : (i += 1) {}
        if (i >= fmt.len) {
            try out.append(allocator, '%');
            try out.appendSlice(allocator, fmt[spec_start..i]);
            break;
        }

        const conv = fmt[i];
        i += 1;
        const arg = argAt(args, ai.*);
        switch (conv) {
            's' => {
                const body = if (precision) |p| arg[0..@min(p, arg.len)] else arg;
                try emitStr(allocator, out, body, width, left);
                ai.* += 1;
            },
            'b' => {
                var decoded = std.ArrayList(u8).empty;
                var j: usize = 0;
                var stop = false;
                while (j < arg.len) {
                    if (arg[j] == '\\') {
                        const escaped = try escapeAt(allocator, arg, j, &decoded);
                        j = escaped.next;
                        if (escaped.stop) {
                            stop = true;
                            break;
                        }
                    } else {
                        try decoded.append(allocator, arg[j]);
                        j += 1;
                    }
                }
                const body = if (precision) |p| decoded.items[0..@min(p, decoded.items.len)] else decoded.items;
                try emitStr(allocator, out, body, width, left);
                ai.* += 1;
                if (stop) return true;
            },
            'c' => {
                const one = arg[0..@min(arg.len, 1)];
                try emitStr(allocator, out, one, width, left);
                ai.* += 1;
            },
            'd', 'i' => {
                const v = argNum(arg, err);
                var digits = std.ArrayList(u8).empty;
                if (!(precision != null and precision.? == 0 and v == 0)) {
                    const mag: u64 = if (v < 0) @intCast(0 -% v) else @intCast(v);
                    try pushRadix(allocator, &digits, mag, 10, false);
                }
                const sign: []const u8 = if (v < 0) "-" else if (plus) "+" else if (space) " " else "";
                try emitNum(allocator, out, sign, digits.items, width, precision, left, zero);
                ai.* += 1;
            },
            'u', 'o', 'x', 'X' => {
                const v: u64 = @bitCast(argNum(arg, err));
                var digits = std.ArrayList(u8).empty;
                const radix: u64 = switch (conv) {
                    'o' => 8,
                    'u' => 10,
                    else => 16,
                };
                if (!(precision != null and precision.? == 0 and v == 0)) {
                    try pushRadix(allocator, &digits, v, radix, conv == 'X');
                }
                var sign = std.ArrayList(u8).empty;
                if (hash and v != 0) {
                    switch (conv) {
                        'o' => if (digits.items.len == 0 or digits.items[0] != '0') try sign.append(allocator, '0'),
                        'x' => try sign.appendSlice(allocator, "0x"),
                        'X' => try sign.appendSlice(allocator, "0X"),
                        else => {},
                    }
                }
                try emitNum(allocator, out, sign.items, digits.items, width, precision, left, zero);
                ai.* += 1;
            },
            else => {
                try out.append(allocator, '%');
                try out.appendSlice(allocator, fmt[spec_start..i]);
            },
        }
    }
    return false;
}

fn argAt(args: []const []const u8, idx: usize) []const u8 {
    return if (idx < args.len) args[idx] else "";
}

fn isLengthModifier(ch: u8) bool {
    return ch == 'l' or ch == 'h' or ch == 'L' or ch == 'q' or ch == 'j' or ch == 'z' or ch == 't';
}

fn pushRadix(allocator: std.mem.Allocator, out: *std.ArrayList(u8), value: u64, radix: u64, upper: bool) !void {
    const lower = "0123456789abcdef";
    const upper_digits = "0123456789ABCDEF";
    const digits = if (upper) upper_digits else lower;
    var tmp: [64]u8 = undefined;
    var i = tmp.len;
    var v = value;
    if (v == 0) {
        i -= 1;
        tmp[i] = '0';
    }
    while (v > 0) {
        i -= 1;
        tmp[i] = digits[@intCast(v % radix)];
        v /= radix;
    }
    try out.appendSlice(allocator, tmp[i..]);
}

const EscapeResult = struct {
    next: usize,
    stop: bool,
};

pub fn escapeAt(allocator: std.mem.Allocator, s: []const u8, i: usize, out: *std.ArrayList(u8)) !EscapeResult {
    if (i + 1 >= s.len) {
        try out.append(allocator, '\\');
        return .{ .next = i + 1, .stop = false };
    }
    const n = s[i + 1];
    switch (n) {
        'a' => return falsePush(allocator, out, i + 2, 7),
        'b' => return falsePush(allocator, out, i + 2, 8),
        'f' => return falsePush(allocator, out, i + 2, 12),
        'n' => return falsePush(allocator, out, i + 2, '\n'),
        'r' => return falsePush(allocator, out, i + 2, '\r'),
        't' => return falsePush(allocator, out, i + 2, '\t'),
        'v' => return falsePush(allocator, out, i + 2, 11),
        '\\' => return falsePush(allocator, out, i + 2, '\\'),
        'c' => return .{ .next = i + 2, .stop = true },
        'x' => return hexEscape(allocator, s, i, out),
        'u', 'U' => return unicodeEscape(allocator, s, i, out, n == 'u'),
        '0' => return octalEscape(allocator, s, i + 2, 3, out),
        '1'...'7' => return octalEscape(allocator, s, i + 1, 3, out),
        else => |other| {
            try out.append(allocator, '\\');
            try out.append(allocator, other);
            return .{ .next = i + 2, .stop = false };
        },
    }
}

fn falsePush(allocator: std.mem.Allocator, out: *std.ArrayList(u8), next: usize, byte: u8) !EscapeResult {
    try out.append(allocator, byte);
    return .{ .next = next, .stop = false };
}

fn hexEscape(allocator: std.mem.Allocator, s: []const u8, i: usize, out: *std.ArrayList(u8)) !EscapeResult {
    var j = i + 2;
    var value: u32 = 0;
    var count: usize = 0;
    while (j < s.len and count < 2) {
        const h = hexVal(s[j]) orelse break;
        value = value * 16 + h;
        j += 1;
        count += 1;
    }
    if (count == 0) {
        try out.append(allocator, '\\');
        try out.append(allocator, 'x');
        return .{ .next = i + 2, .stop = false };
    }
    try out.append(allocator, @truncate(value));
    return .{ .next = j, .stop = false };
}

fn unicodeEscape(allocator: std.mem.Allocator, s: []const u8, i: usize, out: *std.ArrayList(u8), small: bool) !EscapeResult {
    const max: usize = if (small) 4 else 8;
    var j = i + 2;
    var value: u32 = 0;
    var count: usize = 0;
    while (j < s.len and count < max) {
        const h = hexVal(s[j]) orelse break;
        value = value * 16 + h;
        j += 1;
        count += 1;
    }
    if (count == 0) {
        try out.append(allocator, '\\');
        try out.append(allocator, if (small) 'u' else 'U');
        return .{ .next = i + 2, .stop = false };
    }
    try pushUtf8(allocator, out, value);
    return .{ .next = j, .stop = false };
}

fn octalEscape(allocator: std.mem.Allocator, s: []const u8, start: usize, max: usize, out: *std.ArrayList(u8)) !EscapeResult {
    var j = start;
    var value: u32 = 0;
    var count: usize = 0;
    while (j < s.len and count < max and s[j] >= '0' and s[j] <= '7') {
        value = value * 8 + (s[j] - '0');
        j += 1;
        count += 1;
    }
    try out.append(allocator, @truncate(value));
    return .{ .next = j, .stop = false };
}

fn hexVal(ch: u8) ?u32 {
    return switch (ch) {
        '0'...'9' => ch - '0',
        'a'...'f' => ch - 'a' + 10,
        'A'...'F' => ch - 'A' + 10,
        else => null,
    };
}

fn pushUtf8(allocator: std.mem.Allocator, out: *std.ArrayList(u8), cp: u32) !void {
    if (cp < 0x80) {
        try out.append(allocator, @intCast(cp));
    } else if (cp < 0x800) {
        try out.append(allocator, @intCast(0xC0 | (cp >> 6)));
        try out.append(allocator, @intCast(0x80 | (cp & 0x3f)));
    } else if (cp < 0x10000) {
        try out.append(allocator, @intCast(0xE0 | (cp >> 12)));
        try out.append(allocator, @intCast(0x80 | ((cp >> 6) & 0x3f)));
        try out.append(allocator, @intCast(0x80 | (cp & 0x3f)));
    } else {
        try out.append(allocator, @intCast(0xF0 | (cp >> 18)));
        try out.append(allocator, @intCast(0x80 | ((cp >> 12) & 0x3f)));
        try out.append(allocator, @intCast(0x80 | ((cp >> 6) & 0x3f)));
        try out.append(allocator, @intCast(0x80 | (cp & 0x3f)));
    }
}

fn argNum(arg: []const u8, err: *bool) i64 {
    if (arg.len == 0) return 0;
    if (arg[0] == '\'' or arg[0] == '"') {
        return if (arg.len > 1) arg[1] else 0;
    }
    const neg = arg[0] == '-';
    var i: usize = if (neg or arg[0] == '+') 1 else 0;
    const start = i;
    var value: i64 = 0;
    while (i < arg.len and std.ascii.isDigit(arg[i])) : (i += 1) {
        value = std.math.add(i64, std.math.mul(i64, value, 10) catch std.math.maxInt(i64), arg[i] - '0') catch std.math.maxInt(i64);
    }
    if (i == start or i != arg.len) err.* = true;
    return if (neg) -value else value;
}

fn emitNum(
    allocator: std.mem.Allocator,
    out: *std.ArrayList(u8),
    sign: []const u8,
    digits: []const u8,
    width: usize,
    precision: ?usize,
    left: bool,
    zero: bool,
) !void {
    var body = std.ArrayList(u8).empty;
    if (precision) |p| {
        if (p > digits.len) {
            try body.appendNTimes(allocator, '0', p - digits.len);
        }
    }
    try body.appendSlice(allocator, digits);
    const content = sign.len + body.items.len;
    const pad = width -| content;
    if (left) {
        try out.appendSlice(allocator, sign);
        try out.appendSlice(allocator, body.items);
        try out.appendNTimes(allocator, ' ', pad);
    } else if (zero and precision == null) {
        try out.appendSlice(allocator, sign);
        try out.appendNTimes(allocator, '0', pad);
        try out.appendSlice(allocator, body.items);
    } else {
        try out.appendNTimes(allocator, ' ', pad);
        try out.appendSlice(allocator, sign);
        try out.appendSlice(allocator, body.items);
    }
}

fn emitStr(allocator: std.mem.Allocator, out: *std.ArrayList(u8), body: []const u8, width: usize, left: bool) !void {
    const pad = width -| body.len;
    if (left) {
        try out.appendSlice(allocator, body);
        try out.appendNTimes(allocator, ' ', pad);
    } else {
        try out.appendNTimes(allocator, ' ', pad);
        try out.appendSlice(allocator, body);
    }
}
